test/gui/blueprint.lua

```diff
@@ -205,6 +205,41 @@ end
 -- clear and reshow ui if gui/blueprint is run while currently shown
 
 -- mouse support for selecting boundary tiles
+local function click_mouse_and_test(screenx, screeny, should_mark, comment)
+    mock.patch(dfhack.screen, 'getMousePos', mock.func(screenx, screeny),
+        function()
+            local view = load_ui()
+            view:onInput({_MOUSE_L=true})
+            if not should_mark then
+                expect.nil_(view.mark, comment)
+            else
+                local expected_mark = {x=df.global.window_x+screenx-1,
+                                       y=df.global.window_y+screeny-1,
+                                       z=df.global.window_z}
+                expect.table_eq(expected_mark, view.mark, comment)
+                send_keys('LEAVESCREEN') -- cancel selection
+            end
+            send_keys('LEAVESCREEN') -- cancel out of UI
+        end)
+end
+
+function test.set_with_mouse()
+    click_mouse_and_test(0, 0)
+    click_mouse_and_test(0, 5)
+    click_mouse_and_test(5, 0)
+    click_mouse_and_test(5, -1)
+    click_mouse_and_test(-1, 5)
+
+    click_mouse_and_test(5, 7, true, 'interior tile')
+
+    local screen_width, screen_height = dfhack.screen.getWindowSize()
+    click_mouse_and_test(screen_width - (guidm.MENU_WIDTH+3), 7, true,
+                         'just to left of border between map and blueprint gui')
+    click_mouse_and_test(screen_width - (guidm.MENU_WIDTH+2), 7, false,
+                         'on border between map and blueprint gui')
+    click_mouse_and_test(5, screen_height - 2, true, 'above bottom border')
```

Comment:
I can't seem to get the tests with `should_mark=true` to pass locally. Are they working for you? From a bit of further debugging and modifying `expect.*()` to print better tracebacks, it seems that `view.mark` is nil in all cases:

```
Check failed! interior tile: operands to table_eq must be tables or userdata
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:220
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:209
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:234

Check failed! just to left of border between map and blueprint gui: operands to table_eq must be tables or userdata
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:220
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:209
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:237

Check failed! above bottom border: operands to table_eq must be tables or userdata
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:220
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:209
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:241

test failed: gui/blueprint:set_with_mouse
```

Reply:
tests do pass for me. Could it be that `gui.is_in_rect(guidm.getPanelLayout().map, x, y)` is returning false on your map? or is `dfhack.screen.getWindowSize()` returning garbage?

I added the following lines to the test file between the call to `getWindowSize()` and the first call to `click_mouse_and_test()` with `should_mark` set to `true`:
```
    print(('screen_width=%d, screen_height=%d'):format(screen_width, screen_height))
    local screenx, screeny = screen_width - (guidm.MENU_WIDTH+3), 7
    print(('screenx=%d, screeny=%d'):format(screenx, screeny))
    print('map rect:') printall(guidm.getPanelLayout().map)
    print('is_in_rect: '..tostring(gui.is_in_rect(guidm.getPanelLayout().map, screenx, screeny)))
```
and got the following output:
```
screen_width=100, screen_height=63
screenx=67, screeny=7
map rect:
x1                       = 1
y1                       = 1
width                    = 67
height                   = 61
y2                       = 61
x2                       = 67
is_in_rect: true
test passed in 160 ms: gui/blueprint:set_with_mouse
```
If you instrument similarly, what output do you get?

Reply:
I managed to resize the window to the same size as yours, and I get the errors I was getting before. Seems to be consistent (except for the output of this instrumentation) across window sizes, sidebar layouts (pressing `Tab`), and starting modes (Default, LookAround).

Full output:
```
[DFHack]# test -d ../../code/dfhack/scripts/test -m fortress -t set_with_mouse
Loading tests from ../../code/dfhack/scripts/test
Loading file: gui/blueprint.lua
Loading file: quickfort/aliases_unit.lua
Loading file: quickfort/building_unit.lua
Loading file: quickfort/common_unit.lua
Loading file: quickfort/keycodes_unit.lua
Loading file: quickfort/map_unit.lua
Loading file: quickfort/notes_unit.lua
Loading file: quickfort/parse_unit.lua
Loading file: quickfort/reader_unit.lua
Loading file: quickfort/set_unit.lua
Loading file: teleport.lua
Filtering tests by name pattern:
  set_with_mouse
Filtering tests by mode:
  fortress
Selected tests: 1/89
Running 1 tests
Loaded fortress map
Check failed! interior tile: operands to table_eq must be tables or userdata
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:219
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:209
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:233

screen_width=100, screen_height=63
screenx=67, screeny=7
map rect:
x1                       = 1
y1                       = 1
height                   = 61
width                    = 67
y2                       = 61
x2                       = 67
is_in_rect: true
Check failed! just to left of border between map and blueprint gui: operands to table_eq must be tables or userdata
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:219
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:209
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:241

Check failed! above bottom border: operands to table_eq must be tables or userdata
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:219
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:209
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:245

test failed: gui/blueprint:set_with_mouse

Test summary:
0/1 tests passed
7/10 checks passed
0 test files failed to load
0 tests in unreachable modes
```

Reply:
Oh, I'm not testing with https://github.com/DFHack/dfhack/pull/1872 - that's probably causing the patched `getMousePos()` to behave incorrectly. Will try that change.

Reply:
Ok, with that `mock()` change merged, I'm down to one failure. Using https://github.com/DFHack/dfhack/commit/4945018e92ac5caae9a6bc189060f01b092ea1a0 and c453a53bce292caea2e608ac86942dae4d92e73f with this PR merged in.

```
screen_width=80, screen_height=25
screenx=47, screeny=7
map rect:
height                 	 = 23
y2                     	 = 23
x1                     	 = 1
width                  	 = 23
x2                     	 = 23
y1                     	 = 1
is_in_rect: false
Check failed! just to left of border between map and blueprint gui: operands to table_eq must be tables or userdata
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:219
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:209
  at ../../code/dfhack/scripts/test/gui/blueprint.lua:241

test failed: gui/blueprint:set_with_mouse
```

Reply:
> Oh, I'm not testing with [DFHack/dfhack#1872](https://github.com/DFHack/dfhack/pull/1872) - that's probably causing the patched `getMousePos()` to behave incorrectly. Will try that change.

Gah, sorry, I forgot to document that dependency for this PR!

For that last test, what X position is your map border at?

Reply:
Is x2=23 from my output what you're looking for, or something else?

Reply:
Oh, I'm using the default sidebar layout (I think) that has the minimap visible. If I press `Tab` a couple times so that only the sidebar is visible, and only takes up a single pane's worth of space, then the test passes. The failing test is clicking at (47,7), which is definitely in the sidebar before.

I think this can be fixed fairly easily by checking `dfhack.gui.getDwarfmodeViewDims()` - I'll try that on my end.